Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/storage/KeyCacheRepository.kt

```diff
@@ -0,0 +1,151 @@
+import de.rki.coronawarnapp.diagnosiskeys.server.LocationCode
+import de.rki.coronawarnapp.util.TimeStamper
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import org.joda.time.LocalDate
+import org.joda.time.LocalTime
+import timber.log.Timber
+import java.io.File
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class KeyCacheRepository @Inject constructor(
+    private val context: Context,
+    private val databaseFactory: KeyCacheDatabase.Factory,
+    private val timeStamper: TimeStamper
+) {
+
+    private val storageDir by lazy {
+        File(context.cacheDir, "diagnosis_keys").apply {
+            if (!exists()) {
+                if (mkdirs()) {
+                    Timber.d("KeyCache directory created: %s", this)
+                } else {
+                    Timber.w("KeyCache directory creation failed: %s", this)
+                }
+            }
+        }
+    }
+
+    private val database by lazy { databaseFactory.create() }
+
+    private var isInitDone = false
+    private val initMutex = Mutex()
+
+    private suspend fun getDao(): KeyCacheDatabase.CachedKeyFileDao {
+        val dao = database.cachedKeyFiles()
+
+        if (!isInitDone) {
+            initMutex.withLock {
+                if (isInitDone) return@withLock
+                isInitDone = true
+
+                doHouseKeeping()
+            }
+        }
+
+        return dao
+    }
+
+    private suspend fun doHouseKeeping() {
+        val dirtyInfos = getDao().getAllEntries().filter {
+            it.isDownloadComplete && !getPathForKey(it).exists()
+        }
+        delete(dirtyInfos)
+    }
+
+    fun getPathForKey(cachedKeyInfo: CachedKeyInfo): File {
+        return File(storageDir, cachedKeyInfo.fileName)
+    }
+
+    suspend fun getAllCachedKeys(): List<Pair<CachedKeyInfo, File>> {
+        return getDao().getAllEntries().map { it to getPathForKey(it) }
+    }
+
+    suspend fun getEntriesForType(type: CachedKeyInfo.Type): List<Pair<CachedKeyInfo, File>> {
+        return getDao().getEntriesForType(type.typeValue).map { it to getPathForKey(it) }
+    }
+
+    suspend fun createCacheEntry(
+        type: CachedKeyInfo.Type,
+        location: LocationCode,
+        dayIdentifier: LocalDate,
+        hourIdentifier: LocalTime?
+    ): Pair<CachedKeyInfo, File> {
+        val newKeyFile = CachedKeyInfo(
+            type = type,
+            location = location,
+            day = dayIdentifier,
+            hour = hourIdentifier,
+            createdAt = timeStamper.nowUTC
+        )
+
+        val targetFile = getPathForKey(newKeyFile)
+
+        try {
+            getDao().insertEntry(newKeyFile)
+            if (targetFile.exists()) {
+                Timber.w("Target path despire no collision exists, deleting: %s", targetFile)
+            }
+        } catch (e: SQLiteConstraintException) {
+            Timber.e(e, "Insertion collision? Overwriting for %s", newKeyFile)
```

Comment:
AFAIK, this will only be thrown if we violate the uniqueness of the entity ID. Our ID is a stable calculation based on type, location, day (and hour), so trying to insert something with the same ID means that we are trying to re-insert the same day.

`KeyFileDownloader` first creates the entry, then starts the download. If anything goes wrong, we might have an entry and no file. On next download attempt (but not necessarily new app session, where we do house keeping), we might attempt to download the same key again. While this shouldn't happen, and we do log an error here, I think overwriting as recovery is not a bad idea instead of throwing an exception, as we can actually keep going then.

Good think we talked about this, we should also delete the cache entry directly if the download fails, i.e.:

```kotlin
                        return@async try {
                            downloadKeyFile(keyInfo, path)
                            keyInfo to path
                        } catch (e: Exception) {
                            Timber.tag(TAG).e(e, "Download failed: %s", keyInfo)
                            keyCache.delete(listOf(keyInfo))
                            null
                        }

```

Thoughts?